README.md
# HTMLShots
## yet another phantom.js screenshots app

**HTMLShots** is a simple **phantom.js** app that creates screenshots from *HTML* files.
Unlike other similar screenshot solutions, **HTMLShots** makes use of the `WebServer` module
so there is no need to spawn one process for each screenshots request.

### requirements

* PhantomJS >= 1.6.0
* ttf-mscorefonts-installer (optional, only if you want Windows fonts)

### install

#### package

We maintain packages for **Ubuntu 12.04 LTS (precise)** in our PPA:

    $ sudo apt-get install -y python-software-properties
    $ sudo add-apt-repository ppa:templateria/htmlshots
    $ sudo apt-get update
    $ sudo apt-get install htmlshots

**Note:** Ubuntu 12.04 LTS (precise) ships PhantomJS 1.4.0, so our PPA repository provides an updated version (1.6.0),
backported from Ubuntu 12.10 (quantal).

By default the **HTMLShots** service is disabled. To enable it, open */etc/default/htmlshots* and set ENABLED=1, then:

    $ sudo start htmlshots

*Note:* **HTMLShots** supports running multiple instances (processes) with the *INSTANCES* parameter in */etc/default/htmlshots*.
This is particularly useful if you are building a service on top of **HTMLShots** and wants to take advantage of running multiple
processes.

If you want to build the package by yourself:

    $ git clone https://github.com/templateria/htmlshots
    $ cd htmlshots
    $ dpkg-buildpackage

#### git

When installing directly from *git*, first make sure you have installed PhantomJS 1.6.0, then:

    $ git clone https://github.com/templateria/htmlshots

To run HTMLShots on port 3000 (default):

    $ cd htmlshots
    $ phantomjs --config=conf/htmlshots.json htmlshots.js

If you want **HTMLShots** to listen in a different port, pass an additional parameter:

    $ phantomjs --config=conf/htmlshots.json htmlshots.js 8080

### how to use

Note: All screenshots are PNG images returned as base64-encoded strings.

**HTMLShots** is not supposed to be a REST API, just a simple way to retrieve screenshots from HTML files.

Currently, the following methods are supported:

#### PUT

Send a PUT request with the HTML contents in the request body. This example uses cURL's *-T (upload file)* parameter
to simplify things:

    $ curl -T index.html http://localhost:3000/ | base64 -d > screenshot.png

#### POST

Send a POST request with the HTML contents URL-encoded in the *html* parameter.

    $ curl -X POST -d html=contents http://localhost:3000/ | base64 -d > screenshot.png

TODO: Send a POST request with the URL for the desired page in the *url* parameter.

    $ curl http://localhost:3000/?url=http://www.google.com/ | base64 -d > screenshot.png


debian/changelog
htmlshots (0.4.0-1) precise; urgency=low

  * adding suport for rendering a specific element using a selector argument.
  * fixing build instructions in README.md.

 -- Pedro Padron <pedro@templateria.com>  Sat, 11 Jan 2014 06:05:00 -0200

htmlshots (0.3.0-1) precise; urgency=low

  * there is no need for waitFor() anymore, since we are manually setting page.content now
  * added runlevel and respawn to upstart job
  * added README.md and TODO.md
  * added ttf-mscorefonts-installer dependency for better font rendering
  * setting bgColor to white to prevent screenshots with black blackground

 -- W3P Packages Team <packages@w3p.com.br>  Tue, 23 Oct 2012 14:52:20 -0200

htmlshots (0.2.0-1) precise; urgency=low

  * renamed fromfile.js to htmlshots.js

 -- W3P Packages Team <packages@w3p.com.br>  Tue, Oct 2 02:24:31 2012 -0300

htmlshots (0.1.2-1) precise; urgency=low

  * [fix] fixing parent service name in worker upstart script.

 -- W3P Packages Team <packages@w3p.com.br>  Mon, 02 Oct 2012 02:23:52 -0300

htmlshots (0.1.1-1) precise; urgency=low

  * [fix] HTMLShots worker must use the config file in /etc/htmlshots.

 -- W3P Packages Team <packages@w3p.com.br>  Mon, 01 Oct 2012 15:44:00 -0300

htmlshots (0.1.0-1) precise; urgency=low

  * Initial release.

 -- W3P Packages Team <packages@w3p.com.br>  Mon, 01 Oct 2012 13:13:00 -0300


fromfile.js
var port    = phantom.args[0] ? phantom.args[0] : 3000;
var server  = require('webserver').create();
var page    = require('webpage').create();

/**
 * Waits for maxTimeOut until a certain condition is met.
 * 
 * @param  callback testFx     Callback that returns true/false.
 * @param  callback onReady    Callback to be executed when condition is met.
 * @param  callback onTimeout  Callback to be executed when timeout is reached.
 * @param  int      maxTimeOut Max timeout in milliseconds.
 * 
 * @return void
 */
function waitFor(testFx, onReady, onTimeout, maxTimeOut) {
    var maxTimeOut = maxTimeOut ? maxTimeOut : 3000;
    var start      = new Date().getTime();
    var condition  = false;
    var check      = function () {
        if ((new Date().getTime() - start < maxTimeOut) && !condition ) {
            condition = (typeof(testFx) === "string" ? eval(testFx) : testFx());
        } else {
            if(!condition) {
                typeof(onTimeout) === "string" ? eval(onTimeout) : onTimeout();
                clearInterval(interval);
            } else {
                typeof(onReady) === "string" ? eval(onReady) : onReady();
                clearInterval(interval);
            }
        }
    };
    interval = setInterval(check, 400);
}

/**
 * Checks if all images have been loaded.
 * 
 * @return boolean
 */
function imagesLoaded() {
    return page.evaluate(function() {
        var images     = document.getElementsByTagName('img');
        var isComplete = true;
        for (var i = 0; i < images.length; i++) {
            if (!images[i].complete) {
                isComplete = false;
            }
        }
        return isComplete;
    });
}

var service = server.listen(port, function (request, response) {

    var send = function (httpStatus, message) {
        response.statusCode = httpStatus;
        response.write(typeof(message) == 'object' ? JSON.stringify(message) : message);
        return response.close();
    }

    if (request.method != 'PUT' && request.method != 'POST') {
        return send(415, {error: 'Only PUT and POST are allowed.'});
    }

    if (request.method == 'POST' && typeof(request.post) == 'undefined') {
        return send(400, {error: 'Field "html" is required.'});
    }

    var contents = request.method == 'PUT' ? request.post : request.post.html;

    if (!contents) {
        return send(400, {error: 'Content is empty.'});
    }

    page.viewportSize = { width: 1024, height: 768 };
    page.onLoadFinished = function (status) {
        if (status != 'success') {
            send(500, {error: 'Sorry, something wrong happened.'});
        }
        waitFor(imagesLoaded,
            function () { return send(200, page.renderBase64()); },
            function () { return send(500, {error: 'A timeout has ocurred.'}); },
            10000
        );
    };
    page.content = contents;
});


htmlshots.js
var port    = phantom.args[0] ? phantom.args[0] : 3000;
var server  = require('webserver').create();
var page    = require('webpage').create();

var service = server.listen(port, function (request, response) {

    var send = function (httpStatus, message) {
        response.statusCode = httpStatus;
        response.write(typeof(message) == 'object' ? JSON.stringify(message) : message);
        return response.close();
    }

    if (request.method != 'PUT' && request.method != 'POST') {
        return send(415, {error: 'Only PUT and POST are allowed.'});
    }

    if (request.method == 'POST' && typeof(request.post) == 'undefined') {
        return send(400, {error: 'Field "html" is required.'});
    }

    var contents = request.method == 'PUT' ? request.post : request.post.html;

    if (!contents) {
        return send(400, {error: 'Content is empty.'});
    }

    page.viewportSize = { width: 1024, height: 768 };
    page.onLoadFinished = function (status) {
        if (status != 'success') {
            send(500, {error: 'Sorry, something wrong happened.'});
        }
        page.evaluate(function() {
            document.body.bgColor = 'white';
        });

        if (typeof(request.post.selector) != 'undefined') {
            var clipRect = page.evaluate(function (selector) {
                return document.querySelector(selector).getBoundingClientRect();
            }, request.post.selector);
            page.clipRect = {
                top: clipRect.top,
                left: clipRect.left,
                width: clipRect.width,
                height: clipRect.height
            };
        }

        return send(200, page.renderBase64());
    };
    page.content = contents;
});
